### PR TITLE
Implemented margin ranking loss

### PR DESCRIPTION
**Pull request by AndreiMihalea**
Implemented the margin ranking loss as proposed in the issue #2253 I opened. Can also be related to #2200

### CODE REVIEW

**Comment by AndreiMihalea** [comment]
Forgot to pull the latest update. Will reopen when done.

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class MarginRankingLoss
+{
+ public:
+  /**
+   * Create the MarginRankingLoss object with Hyperparameter margin.
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: Hi @AndreiMihalea, Could you please add some details about hyper parameter `margin`. I think you also need to add `@param margin here`. Thanks.

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+   * @param x1 First input data used for evaluating the specified function.
+   * @param x2 Second input data used for evaluating the specified function.
+   * @param y The label vector which contains -1 or 1 values.
+   */
+  template <
+    typename FirstInputType,
+    typename SecondInputType,
+    typename ThirdInputType
+  >
+
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: I don't think we need an empty line here. Could you please remove it. Thanks.

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ */
+#ifndef MLPACK_ANN_LOSS_FUNCTION_MARGIN_RANKING_LOSS_HPP
+#define MLPACK_ANN_LOSS_FUNCTION_MARGIN_RANKING_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template <
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: Hi @AndreiMihalea, Could you also please add template parameters description. They go along the line of : 
```
@tparam InputDataType Type of the input data (arma::colvec, arma::mat, arma::sp_mat or arma::cube).
```
You might need to wrap this in two lines as mlpack limits each line to 80 characters.

File: src/mlpack/tests/loss_functions_test.cpp [hunk 4]
```diff
@@ -510,4 +511,58 @@ BOOST_AUTO_TEST_CASE(LogCoshLossTest)
...
+  // >>> import torch
+  // >>> import torch.nn.functional as F
+  // >>> x1 = torch.tensor([1., 2., 5., 7., -1., -3.])
+  // >>> x2 = torch.tensor([-1., 3., -4., 11., 3., -3.])
+  // >>> y = torch.tensor([1., -1., -1., 1., -1., 1.], requires_grad=True)
+  // >>> loss = F.margin_ranking_loss(x1, x2, y, margin=1.)
+  // >>> loss.item()
+  // 2.6666667461395264
+  // >>> y.grad
+  // tensor([-0.0000,  0.1667, -1.5000,  0.6667,  0.0000, -0.0000])
```
Comment (kartikdutt18) [inline src/mlpack/tests/loss_functions_test.cpp]: Just one more thing, If you want to make change it's totally up to you.
I am not really sure about this comment. I think you could simply write `Calculated using torch.nn.functional.margin_ranking_loss().`
as this comment isn't directly related to mlpack. However I think a member of this organization will provide better insight for this.
Thanks.

Comment (AndreiMihalea) [reply src/mlpack/tests/loss_functions_test.cpp]: Might add it to a Colab notebook for reference. Thanks for all the feedback.

Comment (kartikdutt18) [reply src/mlpack/tests/loss_functions_test.cpp]: Great. That's what I do too.
Thanks for making the changes so quickly. This looks fine to me.

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.)

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]:
```suggestion
```

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+   * @param x1 First input data used for evaluating the specified function.
+   * @param x2 Second input data used for evaluating the specified function.
+   * @param y The label vector which contains -1 or 1 values.
+   */
+  template <
+    typename FirstInputType,
+    typename SecondInputType,
+    typename ThirdInputType
+  >
+  double Forward(const FirstInputType& x1,
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: Do you mind to use the same parameter names as done for the other loss functions, to be consistent across the codebase?

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+   * @param y The label vector which contains -1 or 1 values.
+   * @param output The calculated error.
+   */
+  template <
+    typename FirstInputType,
+    typename SecondInputType,
+    typename ThirdInputType,
+    typename OutputType
+  >
+  void Backward(const FirstInputType& x1,
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: See comment above.

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                const SecondInputType& x2,
+                const ThirdInputType& y,
+                OutputType& output);
+
+  //! Get the output parameter.
+  OutputDataType& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the output parameter.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: Do you mean get the margin parameter?

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp (hunk 5 above)
Comment (AndreiMihalea) [reply src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: Deleted it.

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp (hunk 7 above)
Comment (AndreiMihalea) [reply src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: Modified with input1 / input2 and target.

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp (hunk 8 above)
Comment (AndreiMihalea) [reply src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: Yes. Modified it.

**Comment by kartikdutt18** [comment]
Hey @zoq, As pointed in this comment [here](https://github.com/mlpack/mlpack/pull/2209#discussion_r394052798) having loss functions that take 3 inputs or more can't really be used with sequential unless we add a custom layer for that purpose. I might be missing something, but how would you suggest to add such functions in mlpack. Should we add them as it is?
I would love to get your opinion on this. Thanks a lot.

**Comment by kartikdutt18** [comment]
Hey Andrei, There is a change (maybe), Instead of taking three inputs what do you think about taking two concatenated inputs?
For more details refer to this [link](https://github.com/mlpack/mlpack/pull/2209#discussion_r398266505).

**Comment by AndreiMihalea** [comment]
> Hey Andrei, There is a change (maybe), Instead of taking three inputs what do you think about taking two concatenated inputs?
> For more details refer to this [link](https://github.com/mlpack/mlpack/pull/2209#discussion_r398266505).

Hi Kartik. I will work on that during this weekend.

**Comment by kartikdutt18** [comment]
Hey Andrie, No rush, just wanted your opinion on it. I think @zoq  will provide better insight on how you should proceed, till then I would suggest not making any changes incase the current approach makes more sense.

**Comment by kartikdutt18** [comment]
As discussed on IRC, I think it's okay to move forward with changes suggested (Take concatenated input), I think that makes the most sense.

**Comment by AndreiMihalea** [comment]
> As discussed on IRC, I think it's okay to move forward with changes suggested (Take concatenated input), I think that makes the most sense.

I modified the implementation. Hope it's ok now.

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 9]
```diff
@@ -43,35 +43,30 @@ class MarginRankingLoss
...
   template <
-    typename FirstInputType,
-    typename SecondInputType,
+    typename InputType,
     typename TargetType,
     typename OutputType
   >
-  void Backward(const FirstInputType& input1,
-                const SecondInputType& input2,
+  void Backward(const InputType& input1,
```
Comment (prince776) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: Keep it just `input`. I think you did a typo here.

Comment (AndreiMihalea) [reply src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: Yes, forgot to modify this one. Thank you for the observation.

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  /**
+   * Create the MarginRankingLoss object with Hyperparameter margin.
+   * Hyperparameter margin defines a minimum distance between correctly ranked
+   * samples.
+   */
+  MarginRankingLoss(const double margin = 1.0);
+
+  /**
+   * Computes the Margin Ranking Loss function.
+   * Measures the loss between two intputs and a label with -1 and 1 values.
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]:
```suggestion
   * Measures the loss between two inputs and a label with -1 and 1 values.
```

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+   * a value of 1 in the label means the first input should be ranked higher
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]:
```suggestion
   * A value of 1 in the label means the first input should be ranked higher
```

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+   * and a value of -1 means the second input should be ranked higher.
+   * 
+   * @param input Concatenation of the two inputs for evaluating the specified
+   * function.
+   * @param target The label vector which contains -1 or 1 values.
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]:
```suggestion
   * @param target The label vector which contains values of -1 or 1.
```

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+#ifndef MLPACK_ANN_LOSS_FUNCTION_MARGIN_RANKING_LOSS_HPP
+#define MLPACK_ANN_LOSS_FUNCTION_MARGIN_RANKING_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp]: Could you add a description of what Margin Ranking Loss is?

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss_impl.hpp [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+template<typename InputDataType, typename OutputDataType>
+template <
+    typename InputType,
+    typename TargetType
+>
+double MarginRankingLoss<InputDataType, OutputDataType>::Forward(
+    const InputType& input,
+    const TargetType& target)
+{
+  int input_rows = input.n_rows;
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss_impl.hpp]: As per the style guidelines, we use camel casing for all out variables
```suggestion
  const int inputRows = input.n_rows;
```

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss_impl.hpp [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    typename InputType,
+    typename TargetType,
+    typename OutputType
+>
+void MarginRankingLoss<InputDataType, OutputDataType>::Backward(
+    const InputType& input,
+    const TargetType& target,
+    OutputType& output)
+{
+  int input_rows = input.n_rows;
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss_impl.hpp]: Same as aforementioned

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss_impl.hpp [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+  const InputType& input1 = input.rows(0, input_rows / 2 - 1);
+  const InputType& input2 = input.rows(input_rows / 2, input_rows - 1);
+  output = -target % (input1 - input2) + margin;
+  output.elem(arma::find(output >= 0)).ones();
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/margin_ranking_loss_impl.hpp]: Interesting, didn't know this worked :)

File: src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp (hunk 13 above)
Comment (AndreiMihalea) [reply src/mlpack/methods/ann/loss_functions/margin_ranking_loss.hpp:26]: Added. Thank you for the feedback.

**Comment by kartikdutt18** [comment]
Looks good to me as well. Just one last thing, recently we merged #2339, where we templated the return type instead of directly using float. @AndreiMihalea, do you mind changing to that to conform with the rest of the code base. I will post a message in each function that has anything to do with loss functions. Thanks a lot.

**Comment by kartikdutt18** [comment]
Hey there, This code touches the loss function portion of mlpack's codebase. Recently #2339 was merged where the switch was made to templated return type instead of just using double. So before we merge this, kindly make the changes for the same. I'm pasting this in all relevant PRs. Thanks a lot 👍

**Comment by AndreiMihalea** [comment]
> Looks good to me as well. Just one last thing, recently we merged #2339, where we templated the return type instead of directly using double. @AndreiMihalea, do you mind changing to that to conform with the rest of the code base. I will post a message in each function that has anything to do with loss functions. I hope that's okay. Thanks a lot.

Modified it accordingly. Thanks.

**Comment by kartikdutt18** [comment]
Great, No more comments from my side. Thanks a lot for the awesome work. 👍

**Comment by favre49** [comment]
All the tests are passing, I think we can merge this now. Thanks for keeping up with the repo changes, and thanks @kartikdutt18 for pointing it out.
